File: ruby/lib/ipaddress/prefix32.rb

require_relative 'prefix'
require_relative 'ip_bits'

class IPAddress
  class Prefix32
    def self.from(my, num)
      return Prefix32.create(num)
    end

    def self.create(num)
      if (0 <= num && num <= 32)
        #static _FROM: &'static (Fn(&::prefix::Prefix, usize) -> Result<::prefix::Prefix, String>) =
        # &from
        #static _TO_IP_STR: &'static (Fn(&Vec<u16>) -> String) = &to_ip_str
        ip_bits = IpBits.v4()
        bits = ip_bits.bits
        return Prefix.new({
          num: num,
          ip_bits: ip_bits,
          net_mask: Prefix.new_netmask(num, bits),
          vt_from: -> (a,b) { Prefix32.from(a,b) }
          #vt_to_ip_str: _TO_IP_STR,
        })
      end

      return nil
    end
  end
end

File: go/ipaddress/prefix32.go
package ipaddress

import "fmt"

func Prefix32From(my *Prefix, num uint8) ResultPrefix {
	return Prefix32New(num)
}

///  Gives the prefix in IPv4 dotted decimal format,
///  i.e. the canonical netmask we're all used to
///
///    prefix = IPAddress::prefix::Prefix32.new 24
///
///    prefix.to_ip
///      ///  "255.255.255.0"
///
func Prefix32New(num uint8) ResultPrefix {
	if num <= 32 {
		ipBits := IpBitsV4()
		bits := ipBits.Bits
		tmp := New_netmask(uint8(num), bits)
		return &PrefixOk{&Prefix{
			uint8(num),
			ipBits,
			*tmp,
			Prefix32From,
			//vt_to_ip_str: _TO_IP_STR,
		}}
	}
	tmp := fmt.Sprintf("Prefix must be in range 0..32, got: %d", num)
	return &PrefixError{&tmp}
}
